src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -1964,6 +1972,138 @@ namespace FlipSign {
+            const std::function<void (const std::string &, const Location &)> err) {
+        if (args.size() != 2) {
+            err("Intrinsic FloorDiv function accepts exactly 2 arguments", loc);
+        }
+        ASR::ttype_t *type1 = ASRUtils::expr_type(args[0]);
+        ASR::ttype_t *type2 = ASRUtils::expr_type(args[1]);
+        type1 = ASRUtils::type_get_past_const(type1);
+        type2 = ASRUtils::type_get_past_const(type2);
+        if (!((ASRUtils::is_integer(*type1) && ASRUtils::is_integer(*type2)) ||
+            (ASRUtils::is_unsigned_integer(*type1) && ASRUtils::is_unsigned_integer(*type2)) ||
+            (ASRUtils::is_real(*type1) && ASRUtils::is_real(*type2)) ||
+            (ASRUtils::is_logical(*type1) && ASRUtils::is_logical(*type2)))) {
+            err("Argument of the FloorDiv function must be either Real, Integer, Unsigned Integer or Logical",
+                args[0]->base.loc);
+        }
+        ASR::expr_t *m_value = nullptr;
+        double compile_time_arg2_val;
+        if (ASRUtils::extract_value(expr_value(args[1]), compile_time_arg2_val)) {
+            if (compile_time_arg2_val == 0.0) {
+                err("Division by 0 is not allowed", args[1]->base.loc);
+            }
+        }
+        if (all_args_evaluated(args)) {
+            Vec<ASR::expr_t*> arg_values; arg_values.reserve(al, 2);
+            arg_values.push_back(al, expr_value(args[0]));
+            arg_values.push_back(al, expr_value(args[1]));
+            m_value = eval_FloorDiv(al, loc, expr_type(args[1]), arg_values);
+        }
+        return ASR::make_IntrinsicScalarFunction_t(al, loc,
+            static_cast<int64_t>(IntrinsicScalarFunctions::FloorDiv),
+            args.p, args.n, 0, ASRUtils::expr_type(args[0]), m_value);
+    }
+
+    static inline ASR::expr_t* instantiate_FloorDiv(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_optimization_floordiv_" + type_to_str_python(arg_types[1]));
+        fill_func_arg("a", arg_types[0]);
+        fill_func_arg("b", arg_types[1]);
+        auto r = declare("r", real64, Local);
+        auto tmp = declare("tmp", int64, Local);
+        auto result = declare("result", return_type, ReturnVar);
+        /*
+        @overload
+        def _lpython_floordiv(a: i32, b: i32) -> i32:
+            r: f64 # f32 rounds things up and gives incorrect tmps
+            tmp: i64
+            result: i32
+            r = float(a)/float(b)
+            tmp = i64(r)
+            if r < 0.0 and f64(tmp) != r:
+                tmp = tmp - 1
+            result = i32(tmp)
+            return result
+        */
+
+        ASR::expr_t *op1 = r64Div(CastingUtil::perform_casting(args[0], arg_types[0], real64, al, loc),
+            CastingUtil::perform_casting(args[1], arg_types[1], real64, al, loc));
```

Comment:
I can see it is being used here. What kind of input arguments are allowed here? We need to be careful that we do not accept too many inputs, we should be explicit about exactly what types we accept in `_lpython_floordiv`.

Reply:
We currently support `integers`, `unsigned integers`, `reals` and `logicals`.